compiler/rustc_resolve/src/lib.rs

```diff
@@ -1385,6 +1382,19 @@ impl<'tcx> Resolver<'_, 'tcx> {
     pub fn tcx(&self) -> TyCtxt<'tcx> {
         self.tcx
     }
+
+    /// This function is very slow, as it iterates over the entire
+    /// [Resolver::node_id_to_def_id] map just to find the [NodeId]
+    /// that corresponds to the given [LocalDefId]. Only use this in
+    /// diagnostics code paths.
```

Comment:
Perhaps use `set_must_produce_diag` to make sure it's used only for diagnostics?

Reply:
Turns out this is used for UNUSED_MACRO_RULES which is allow-by-default (and sometimes explicitly allowed e.g. in stdarch).